File: server/internal/usecase/monthly_balance.go
package usecase

import (
	"context"

	"github.com/kotarowakuwaku/asset-wish/server/internal/domain"
)

type MonthlyBalanceUsecase struct {
	balances MonthlyBalanceRepository
	newID    IDGenerator
}

func NewMonthlyBalanceUsecase(balances MonthlyBalanceRepository, newID IDGenerator) *MonthlyBalanceUsecase {
	return &MonthlyBalanceUsecase{balances: balances, newID: newID}
}

// List は年月の降順で全件返す。
func (u *MonthlyBalanceUsecase) List(ctx context.Context) ([]domain.MonthlyBalance, error) {
	return u.balances.ListAll(ctx)
}

// Upsert は同じ年月への再登録を上書きとして扱う（冪等）。
//
// 戻り値は保存後の姿。既存の月を上書きした場合、ID は既存行のものに
// なるため、採番した ID をそのまま返すと DB に無い ID を返すことになる。
func (u *MonthlyBalanceUsecase) Upsert(
	ctx context.Context,
	ym domain.YearMonth,
	income, expense domain.Money,
) (domain.MonthlyBalance, error) {
	m, err := domain.NewMonthlyBalance(u.newID(), ym, income, expense)
	if err != nil {
		return domain.MonthlyBalance{}, err
	}
	return u.balances.Upsert(ctx, m)
}

File: worker/src/usecase/monthlyBalance.ts
import type { Money } from '../domain/money'
import { MonthlyBalance } from '../domain/monthlyBalance'
import type { YearMonth } from '../domain/yearMonth'
import type { IDGenerator, MonthlyBalanceRepository } from './port'

export class MonthlyBalanceUsecase {
  readonly #balances: MonthlyBalanceRepository
  readonly #newID: IDGenerator

  constructor(balances: MonthlyBalanceRepository, newID: IDGenerator) {
    this.#balances = balances
    this.#newID = newID
  }

  /** 年月の降順で全件返す。 */
  list(): Promise<MonthlyBalance[]> {
    return this.#balances.listAll()
  }

  /**
   * 同じ年月への再登録を上書きとして扱う（冪等）。
   *
   * 戻り値は保存後の姿。既存の月を上書きした場合、ID は既存行のものになるため、
   * 採番した ID をそのまま返すと DB に無い ID を返すことになる。
   */
  // async にしているのは、検証エラーを同期の throw ではなく reject にするため。
  // 呼び出し側が Promise を受け取る前に投げられると、await で捕まえられない。
  async upsert(yearMonth: YearMonth, income: Money, expense: Money): Promise<MonthlyBalance> {
    const m = MonthlyBalance.create(this.#newID(), yearMonth, income, expense)
    return this.#balances.upsert(m)
  }
}
